Sensible Living protocol added.

Exscotticus opened the pull request:
Protocol for [these cool sensors](https://www.tindie.com/products/Michael_Phillips/mini-plant-moisture-sensor/?pt=ac_prod_search).

merbanan commented:
Can you submit some signal samples? And would it be possible to merge the protocols?

zuckschwerdt commented on include/rtl_433_devices.h:
```diff
@@ -105,7 +105,8 @@
 		DECL(interlogix) \
 		DECL(dish_remote_6_3) \
-		DECL(ss_sensor)
+		DECL(ss_sensor) \
```
use the same indent (tabs)

zuckschwerdt commented on src/CMakeLists.txt:
```diff
@@ -115,6 +115,7 @@ add_executable(rtl_433
 	devices/interlogix.c
 	devices/dish_remote_6_3.c
 	devices/simplisafe.c
+        devices/sensible_living.c
```
indent

zuckschwerdt commented on src/devices/sensible_living.c:
```diff
@@ -0,0 +1,183 @@
+// This is pretty arbitrary
+#define SL_ASK_MAX_PAYLOAD_LEN 80
+#define SL_ASK_HEADER_LEN 4
+#ifndef SL_ASK_MAX_MESSAGE_LEN
```
why would this already be defined?

zuckschwerdt commented on src/devices/sensible_living.c:
```diff
@@ -0,0 +1,183 @@
+static uint8_t symbol_6to4(uint8_t symbol)
+{
+    uint8_t i;
+    // Linear search :-( Could have a 64 byte reverse lookup table?
```
might be worth to try a switch-case and see if the compiler can optimize it

zuckschwerdt commented on src/devices/sensible_living.c:
```diff
@@ -0,0 +1,183 @@
+    // Get time
+    char time_str[LOCAL_TIME_BUFLEN];
+    data_t *data;
+    local_time_str(0, time_str);
```
only call local_time_str just before data_make

zuckschwerdt commented on src/devices/sensible_living.c:
```diff
@@ -0,0 +1,183 @@
+}
+
+
+static int sensible_living_callback(bitbuffer_t *bitbuffer) {
```
this brace `{ ` should have it's own line

zuckschwerdt commented on src/devices/sensible_living.c:
```diff
@@ -0,0 +1,183 @@
+    uint8_t init_pattern_len = 40;
+
+    pos = bitbuffer_search(bitbuffer, row, 0, init_pattern, init_pattern_len);
+    if(pos == len){
```
space after keywords

zuckschwerdt commented on src/devices/sensible_living.c:
```diff
@@ -0,0 +1,183 @@
+    // Get header
+    data_len = msg_len - SL_ASK_HEADER_LEN - 3;
+    house_id = payload[1];
+    module_id = payload[2] * 256 + payload[3];
```
this should probably be a `<<8` and `|`?

zuckschwerdt commented on src/devices/sensible_living.c:
```diff
@@ -0,0 +1,183 @@
+	);
+    data_acquired_handler(data);
+
+    return 0;
```
return `1` on success

zuckschwerdt commented on src/devices/sensible_living.c:
```diff
@@ -0,0 +1,183 @@
+r_device sensible_living = {
+    .name           = "Sensible Living Mini-Plant Moisture Sensor",
+    .modulation     = OOK_PULSE_PCM_RZ,
+    .short_limit    = SL_ASK_BIT_LEN,
```
Please use absolute values. It easier to graps when browsing through the source.

Exscotticus commented:
> Can you submit some signal samples?

Yes I can. Put them in the tests subfolder in the sensible-living branch?

>  And would it be possible to merge the protocols?

Good question. It certainly would be nice to have a single protocol handle all VirtualWire and RadioHead protocols. The first hurdle is that the baud rate is arbitrary. The next hurdle is that some values get split across an arbitrary number of bytes. How would the code know which bytes to combine?

I'll work on your edits. Just keep in mind that the original **radiohead_ask.c** that I modified has the same grammar issues that you want me to address—and you accepted it. I really tried to keep my file the same and only change what was necessary. This makes diffing the two files to see what I changed easier.

zuckschwerdt commented:
oh, hadn't noticed that [radiohead_ask.c](https://github.com/merbanan/rtl_433/blob/master/src/devices/radiohead_ask.c) is in that bad shape. I'll push some style updates.

Exscotticus commented:
Admittedly my code is not good. For example, I'm grabbing 10 bytes from the payload, but I don't even validate if the payload has 10 bytes. I suppose now is a good time to mention that I'm not a C developer. Point is, no offense taken. I'll make the edits.

zuckschwerdt commented:
I have update the code style in radiohead_ask.c somewhat. You should be able to add your device in the same file and that way share some code (the 6to4 stuff).

Add the test samples to the [rtl_433_tests](https://github.com/merbanan/rtl_433_tests/tree/master/tests/radiohead_ask) repo in tests/radiohead_ask/03

zuckschwerdt commented:
You can just keep adding changes to this PR freely. We'll squash the set of commits on merge.

Exscotticus commented:
Added test files [here](https://github.com/Exscotticus/rtl_433/tree/sensible-living/tests/radiohead_ask/03). Does it make sense for me to complete the requested edits to **sensible_living.c** if you now want to try using **radiohead_ask.c** for multiple devices? How will we get around the aforementioned baud/byte challenges? If you're trying to create an abstraction that all VirtualWire/RadioHead can use, then **radiohead_ask.c** shouldn't make any assumptions about the baud rate or the payload.

Exscotticus commented:
> The sample files should go to the other repo, as new PR.

Confused again. You said...

> You can just keep adding changes to this PR freely.

But a PR is associated with a branch, and so I added the test files to the branch associated with this PR. Do you want me to create a new branch and a new PR wherein I only modify **radiohead_ask.c**? If not, please by very explicit. Assume that I have about 24 hours of git training. What repo? What branch?

> See e.g. lacrosse_tx35.c for an example of multiple devices in one source file.

I see it. And I see how a distinction is being made based on different pulse characteristics. But in the case of all VirtualWire/RadioHead devices, the pulse characteristics will be the same. You could distinguish based on bit rate—until some other VirtualWire/RadioHead device comes along with the same bit rate. Then what? Because then the payloads won't be right. You can't distinguish based on payload length, because two devices could have the same length but interpret the bytes differently.

I understand that you're trying to abstract two devices under one protocol. I juts don't understand how you're going to distinguish the umpteen million potential VirtualWire/RadioHead devices in this manner.

zuckschwerdt commented:
The code change and the sample files are not related and placed in two separate Git repositories.

For the samples:
- Clone the rtl_433_tests repository. https://github.com/merbanan/rtl_433_tests/
- Add a branch in your copy
- Add the sample files, Commit, Push
- Pull request

For the code changes keep using this PR #742 (and branch).

If you look in the https://github.com/merbanan/rtl_433/blob/master/src/devices/lacrosse_tx35.c#L184 there are two devices, lacrosse_tx29 and lacrosse_tx35. A device is a protocol. The two protocol/devices just share the same source file, because they are somewhat related.

- Remove the samples from this PR
- Add your r_device and callback from sensible_living.c to radiohead_ask.c

Exscotticus commented:
> Add your r_device and callback from sensible_living.c to radiohead_ask.c

Done.

Exscotticus commented:
Is there anything else this pull request is waiting on?

zuckschwerdt commented:
I hope to get some time in the next days to integrate this. I want to layout the code so it is a bit more reusable.

zuckschwerdt commented:
Have a look at https://github.com/merbanan/rtl_433/blob/master/src/devices/radiohead_ask.c -- it should be easier to understand and extend in the future.

Exscotticus commented:
I agree. Good changes. You abstracted a large chunk of complex common code from both callbacks. Thanks!
